subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/EncryptionService.kt

```diff
@@ -57,49 +61,67 @@ interface EncryptionService : EncryptionConfiguration {
 
 @ServiceScope(Scopes.BuildTree::class)
 internal
-class DefaultEncryptionService(startParameter: ConfigurationCacheStartParameter) : EncryptionService {
-
-    private
-    val encryptingRequested: Boolean = startParameter.encryptionRequested
+class DefaultEncryptionService(
+    startParameter: ConfigurationCacheStartParameter,
+    cacheBuilderFactory: GlobalScopedCacheBuilderFactory,
+) : EncryptionService {
 
     private
     val secretKey: SecretKey? by lazy {
-        if (keySource == null) {
-            null
-        } else
-            try {
-                keySource!!.getKey().also {
-                    val keyLength = it.encoded.size
-                    if (keyLength < 16) {
-                        throw InvalidKeyException("Encryption key length is $keyLength bytes, but must be at least 16 bytes long")
+        when {
+            // encryption is on by default
+            startParameter.encryptionRequested -> {
+                secretKeySourceFrom(startParameter).let { keySource ->
+                    try {
+                        secretKeyFrom(keySource, cacheBuilderFactory).also { key ->
+                            val keyLength = key.encoded.size
+                            if (keyLength < 16) {
+                                throw InvalidKeyException("Encryption key length is $keyLength bytes, but must be at least 16 bytes long")
+                            }
+                            // acquire a cipher just to validate the key
+                            encryptionAlgorithm.newSession(key).encryptingCipher {}
+                        }
+                    } catch (e: EncryptionException) {
+                        if (e.message != null) {
+                            throw e
+                        }
+                        throw EncryptionException("Error loading encryption key from ${keySource.sourceDescription}", e.cause)
+                    } catch (e: Exception) {
+                        throw EncryptionException("Error loading encryption key from ${keySource.sourceDescription}", e)
                     }
-                    // acquire a cipher just to validate the key
-                    encryptionAlgorithm.newSession(it).encryptingCipher {}
                 }
-            } catch (e: EncryptionException) {
-                logger.warn("Encryption was requested but could not be enabled: ${e.message}")
-                null
-            } catch (e: Exception) {
-                logger.warn("Encryption was requested but could not be enabled: $e")
+            }
+
+            else -> {
                 null
             }
+        }
     }
 
+    private
+    fun secretKeyFrom(keySource: KeyStoreKeySource, cacheBuilderFactory: GlobalScopedCacheBuilderFactory) =
+        // JKS does not support non-PrivateKeys
```

Comment:
This comment could be near `loadOrCreateKey` or the `KEYSTORE_TYPE` constant.

Reply:
Done, thanks.